src/Aabb.h

```diff
@@ -42,12 +62,71 @@ struct Aabb {
 	}
 	bool Intersects(const Aabb &o) const
 	{
-		return (min.x < o.max.x) && (max.x > o.min.x) &&
-			(min.y < o.max.y) && (max.y > o.min.y) &&
-			(min.z < o.max.z) && (max.z > o.min.z);
+		return min < o.max && max > o.min;
 	}
 	// returns maximum point radius, usually smaller than max radius of bounding box
 	double GetRadius() const { return radius; }
 };
 
+// SIMD-capable AABB type without additional overhead
+template<typename Vec3>
+struct AABB {
+	using Number = typename Vec3::element_type;
+
+	Vec3 min, max;
+
+	static AABB Invalid()
+	{
+		return AABB{ Vec3(DBL_MAX, DBL_MAX, DBL_MAX), Vec3(-DBL_MAX, -DBL_MAX, -DBL_MAX) };
+	}
+
+	void Update(const AABB &rhs)
+	{
+		using std::max;
+		using std::min;
+
+		this->max.x = max(this->max.x, rhs.max.x);
+		this->max.y = max(this->max.y, rhs.max.y);
+		this->max.z = max(this->max.z, rhs.max.z);
+		this->min.x = min(this->min.x, rhs.min.x);
+		this->min.y = min(this->min.y, rhs.min.y);
+		this->min.z = min(this->min.z, rhs.min.z);
+	}
+
+	void Update(const Vec3 &point)
+	{
+		using std::max;
+		using std::min;
+
+		this->max.x = max(this->max.x, point.x);
+		this->max.y = max(this->max.y, point.y);
+		this->max.z = max(this->max.z, point.z);
+		this->min.x = min(this->min.x, point.x);
+		this->min.y = min(this->min.y, point.y);
+		this->min.z = min(this->min.z, point.z);
+	}
+
+	auto IntersectsRay(const Vec3 &start, const Vec3 &inv_dir, Number dist) const
+	{
+		using std::min;
+		using std::max;
+
+		Vec3 l1 = (this->min - start) * inv_dir;
+		Vec3 l2 = (this->max - start) * inv_dir;
+
+		Vec3 vmin = min(l1, l2);
+		Vec3 vmax = max(l1, l2);
+
+		double lmin = max(vmin.x, max(vmin.y, vmin.z));
+		double lmax = min(vmax.x, min(vmax.y, vmax.z));
+
+		return ((lmax >= 0.f) & (lmax >= lmin) & (lmin < dist));
+	}
+
+	auto IsIn(Vec3 point) const { return point >= min && point <= max; }
+	auto Intersects(const AABB &rhs) const { return min < rhs.max && max > rhs.min; }
+};
```

Comment:
This type confuses me a little.  We have some things using a `min()` `max()` method that takes two vectors (inside `IntersectsRay`) and other things doing component wise min and max.

Then our `Invalid` method seems to have determined that we're using doubles as our component type for our vectors and yet `0.f` inside our `IntersectsRay` method suggests floats.

Is the intention to proved a SIMD optimized variant of Vector3 that has correct alignment constraints and operator overload functionality?

Reply:
The intention for this specific class is to provide a templated, optimized version of the existing Axis-Aligned-Bounding-Box implementation (`Aabb`) which can operate semi-generically on double-precision, single-precision, or packed SIMD vectors, by changing the `Vec3` type provided. E.g. `AABB<vector3<xsimd::batch<float, AVX2>>>` (not exact code) would be an AABB storing eight single-precision float elements "per component" with no extra code compared to the scalar version - it's the responsibility of the calling code to unpack individual elements back to scalar datatypes if necessary.

> Then our Invalid method seems to have determined that we're using doubles as our component type for our vectors and yet 0.f inside our IntersectsRay method suggests floats.

This is a copy-paste error - it should be `Number(0)` instead.

> We have some things using a min() max() method that takes two vectors (inside IntersectsRay) and other things doing component wise min and max.

The min/max on vectors is just a nicer interface compared to manually writing `std::min(a.x, b.x)` etc. everywhere the operation is performed. It uses the same semantics as packed SIMD min/max operations, so it operates per-element and per-component. The component-wise min/max you're seeing in `IntersectsRay` is a "horizontal min/max" (taking the minimum of all elements in a *single* vector) and that's the most optimized way to express it.

...that being said, I see now what you're talking about - I forgot to update the other component-wise calls when I added the min/max overload for our `vector3` type. I also forgot to make the min/max overloads generic enough to use with a vectorized `element_type`. Will fix.

Reply:
Ahh so there is a min/max that returns a scaler of the min/max of a single vector's components and then component-wise min/max too.

I believe most SIMD implementations should accelerate at least the component wise ones to make them single instructions, I'm not sure if we write them component wise though, the compiler will figure that out for us.